pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,551 @@
+#pragma once
+
+#include <bit>
+#include <complex>
+#include <cstring>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <mpi.h>
+#include <stdexcept>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <vector>
+
+#include "Error.hpp"
+
+namespace Pennylane::MPI {
+inline void errhandler(int errcode, const char *str) {
+    char msg[MPI_MAX_ERROR_STRING];
+    int resultlen;
+    MPI_Error_string(errcode, msg, &resultlen);
+    fprintf(stderr, "%s: %s\n", str, msg);
+    MPI_Abort(MPI_COMM_WORLD, 1);
+}
+
+#define PL_MPI_IS_SUCCESS(fn)                                                  \
+    {                                                                          \
+        int errcode;                                                           \
+        errcode = (fn);                                                        \
+        if (errcode != MPI_SUCCESS)                                            \
+            errhandler(errcode, #fn);                                          \
+    }
+
+#define CPPTYPE(s) std::type_index(typeid(s))
```

Comment:
Can this be implemented with a templated function, lambda or another way that does not use the preprocessor?

Reply:
done with a templated function.